### ISSUE TITLE
KeyTables does not function when thead is hidden

### ISSUE THREAD

**Issue by benbankes** [DataTables/DataTables#921]
```
thead {
    display: none;
}
```
Causes no key events to be issued, and no UI changes normal to the KeyTables extension appear.

**Comment by DataTables** [comment]
That's certainly unexpected. Trivial test case here: http://live.datatables.net/zocesivi/1/edit . Not immediately sure what is causing that!

**Comment by benbankes** [comment]
Allan,

Thanks for being quick with the response. A simple workaround is to set a
large, negative "margin-top" on these.

Thanks,
Ben

On Mar 24, 2017 2:23 PM, "Allan Jardine" <notifications@github.com> wrote:

That's certainly unexpected. Trivial test case here:
http://live.datatables.net/zocesivi/1/edit . Not immediately sure what is
causing that!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/DataTables/DataTables/issues/921#issuecomment-289133689>,
or mute the thread

**Comment by DataTables** [comment]
Its caused by [this](https://github.com/DataTables/KeyTable/blob/e43ea0a319aae89388c177d0363f36238063595a/js/dataTables.keyTable.js#L337).

The way Responsive works is to set the column to `display:none`, hence why that function exists. And DataTables uses the header row for the column selectors. So since none are visible here, KeyTable decides that the column can't be focused on.

I'm not actually certain what the correct "fix" is here (I quote fix as I've assumed in the code base in a number of places that the header is required).

**Comment by benbankes** [comment]
Allan,

Interesting.  It might be better to look a the first row of data instead of
the thead, but this probably has some other gotchas of which I am unaware.

Thanks for looking into it!
Ben

**Comment by DataTables** [comment]
The gotcha there is that the table might have no rows to display (zero data, or all filtered out).

**Comment by benbankes** [comment]
Allan,

In that case, I would think you wouldn't need to use the keys plugin,
because there are no columns to be navigated.  However, there are probably
some cases where data is not loaded yet, like when serverSide is true.
Maybe that's what you're referring to.

Thanks,
Ben

**Comment by DataTables** [comment]
Yeah - the problem is that the table can move in and out of that state. There is no way to know in advance if the table might have zero records (for example a new table, that you want to use KeyTable on, but haven't yet entered any data), or filtering has been applied to a table you want to use KeyTable on.

Its an interesting one this one. The only option I've come up with so far is to check the `display` property of each column header, rather than simply using `:visible` as a jQuery selector. But that's going to be slower and involve more code.
